### PR TITLE
Address Eric Crockett's review comments.

### PR DESCRIPTION
**Pull request by chris-wood**
cc @crockeea

### CODE REVIEW

**Comment by chris-wood** [comment]
@hugokraw this PR should be good to go!

**Review by hugokraw — CHANGES_REQUESTED** [review]

**Comment by hugokraw** [comment]
I am afraid that I did not pushed  my comments (which I wrote them a couple of days ago) before for you to see.  Sorry for that. Eric also asked for a couple of clarifications to add (on the need of forward secrecy and something else). If you need text from me let me know.

File: draft-irtf-cfrg-opaque.md [hunk 1]
```diff
@@ -350,8 +354,9 @@ OPAQUE relies on the following protocols and primitives:
     `params` to strengthen the input `msg` against offline dictionary attacks.
     This function also needs to satisfy collision resistance.
 
-We also assume the existence of a function `KeyGen`, which generates an OPRF private
-and public key. We write `(skU, pkU) = KeyGen()` to denote this function.
+We also assume the existence of a function `KeyGen` from {{I-D.irtf-cfrg-voprf}}, which
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply draft-irtf-cfrg-opaque.md:358]: It's currently not part of the OPRF API, so I left it off the list above. If the OPRF draft changes, we can change this too.

File: draft-irtf-cfrg-opaque.md [hunk 2]
```diff
@@ -362,19 +367,13 @@ them as input to an authenticated key exchange (AKE) protocol.
...
-discusses integration with TLS 1.3 {{RFC8446}}.
+key-exchange protocol (AKE) using credentials recovered after the OPRF protocol completes.
+(The key-exchange protocol MUST satisfy the KCI requirement discussed in {{security-considerations}}.)
 
-The instantiations that follow transmit IdU in cleartext. Applications that require
-IdU privacy should encrypt this using appropriately. Mechanisms for doing so are
-outside the scope of this document, though may be addressed elsewhere, such as
-in {{I-D.sullivan-tls-opaque}}.
+We first define the core OPAQUE protocol based on a generic OPRF, hash, and MHF function.
+Specification of the AKE is out of scope for this document. {{instantiations}} describes
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply draft-irtf-cfrg-opaque.md:376]: What I meant is that this document doesn't intend to specify, describe, etc *new* AKEs, but I see how this is confusing. I'll just remove it.

File: draft-irtf-cfrg-opaque.md [hunk 3]
```diff
@@ -876,8 +875,8 @@ Steps:
 [[RFC editor: please change "RFCXXXX" to the correct number before publication.]]
 
 As in the registration phase, applications MUST authenticate pkS; secrecy of pkS is
-optional. If an application requires secrecy of pkS, this value SHOULD be omitted
-from auth_data (step 9).
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply draft-irtf-cfrg-opaque.md:881]: I think this entire paragraph is probably obsolete. I'll just remove it.

File: draft-irtf-cfrg-opaque.md [hunk 4]
```diff
@@ -898,37 +897,37 @@ OPRF computation and credential fetch flow.
...
 
-# Key Exchange Protocol Instantiations {#instantiations}
+# Authenticated Key Exchange Protocol Instantiations {#instantiations}
 
-This section describes several instantiations of OPAQUE using different KE protocols.
-For the sake of concreteness it only includes KE protocols consisting of three messages,
-denoted KE1, KE2, KE3, where KE1 and KE2 include DH values sent by client and
-server, respectively, and KE3 provides explicit client authentication and full
+This section describes several instantiations of OPAQUE using different AKE protocols.
+As discussed in {{security-considerations}}, each AKE protocol requires KCI resistance.
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply draft-irtf-cfrg-opaque.md:922]: That's a nice rewrite -- I'll take it!

File: draft-irtf-cfrg-opaque.md [hunk 5]
```diff
@@ -688,8 +686,9 @@ is that which is associated with the OPAQUE configuration (see {{configurations}
 
 All `CredentialExtension` values require authentication. Only skU requires secrecy.
 If an application requires secrecy of pkS, this value SHOULD be included in the
-`Credentials.secret_credentials` list (step 5). Applications may optionally include
-pkU, IdU, or IdS in the `Credentials.secret_credentials` structure (step 5) if secrecy
+`Credentials.secret_credentials` list (step 5), and MUST NOT be included in the
+`Credentials.cleartext_credentials` list. Applications may optionally include
+pkU, idU, or IdS in the `Credentials.secret_credentials` structure (step 5) if secrecy
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply draft-irtf-cfrg-opaque.md]: Fixed!

File: draft-irtf-cfrg-opaque.md [hunk 6]
```diff
@@ -748,10 +747,10 @@ the OPRF protocol.
 
 Note also that the authenticate stage can run the OPRF and AKE protocols concurrently
 with interleaved and combined messages (while preserving the internal ordering of
```
[the remark this answers is not in the document]
Comment (chris-wood) [reply draft-irtf-cfrg-opaque.md]: Fixed!

**Comment by chris-wood** [comment]
@hugokraw Thanks for the feedback -- I incorporated your comments. I also added some additional changes on top. Namely:

- Fix the auth_tag size to the output of HMAC. (One less variable-length thing to worry about.)
- Change how the auth_tag is computed. Previously, it did not include the length of the ciphertext or auth_data, which was worrisome. Now it's computed over the structure:

```
struct {
  opaque nonce[32];
  opaque ct<1..2^16-1>;
  opaque auth_data<0..2^16-1>;
} InnerEnvelope;
```

which includes the length of `ct` and `auth_data`.

Please let me know what you think!

**Comment by chris-wood** [comment]
@hugokraw, friendly bump! Please feel free to edit the PR directly if that's easier for you. :-)

File: draft-irtf-cfrg-opaque.md [hunk 7]
```diff
@@ -1406,60 +1400,60 @@ whether a given user identity is registered with a server. Preventing
...
-disclose their password to the server, even during registration. Note that this
-does not prevent a malicious server from conducting a dictionary attack on inputs
-provided by the client. OPAQUE assumes the server is honest, and only guarantees
-safeguards against parties who may later compromise the server and any stored
-user account information.
+disclose their password to the server, even during registration. Note that a corrupted
+server can run an exhaustive offline dictionary attack to validate guesses for the user's
+password; this is inevitable in any aPAKE protocol. (OPAQUE enables a defense against such
+offline dictionary attacks by distributing the server so that an offline attack is only
+possible if all - or a minimal number of - servers are compromised.)
```
Comment (hugokraw) [inline draft-irtf-cfrg-opaque.md:1456]: Add a reference to the OPAQUE paper at the end of the parentheses

**Review by hugokraw — APPROVED** [review]

**Comment by hugokraw** [comment]
Sorry for the delay in responding to the latest changes. I was busy welcoming my second grandson... :-)

**Pull request merged**